From 1cf6af3a11ae83b00ce4e3ab8c9be31a076977ad Mon Sep 17 00:00:00 2001
From: Tony Liu <Bo.Liu@windriver.com>
Date: Tue, 30 Mar 2010 15:04:46 +0800
Subject: [PATCH 04/13] lsi_app: add interrupt definition and handling

There is a ARM Generic Interrupt Controller on LSI app3300, merge the platform
IRQ definitions, init and low level serve routines from vendor drop -- Agere APP
Linux Version 2.6.1.5.pre10 7 May 2009.

Signed-off-by: Tonyliu <Bo.Liu@windriver.com>
---
 arch/arm/mach-app/include/mach/entry-macro.S |   70 ++++++++++++++++++
 arch/arm/mach-app/include/mach/irqs.h        |  101 ++++++++++++++++++++++++++
 arch/arm/mach-app/irq.c                      |   29 ++++++++
 3 files changed, 200 insertions(+), 0 deletions(-)
 create mode 100644 arch/arm/mach-app/include/mach/entry-macro.S
 create mode 100644 arch/arm/mach-app/include/mach/irqs.h
 create mode 100644 arch/arm/mach-app/irq.c

diff --git a/arch/arm/mach-app/include/mach/entry-macro.S b/arch/arm/mach-app/include/mach/entry-macro.S
new file mode 100644
index 0000000..9e35934
--- /dev/null
+++ b/arch/arm/mach-app/include/mach/entry-macro.S
@@ -0,0 +1,70 @@
+/*
+ * arch/arm/mach-app/include/mach/entry-macro.S
+ *
+ * Low-level IRQ helper macros for APP3/APP3K platforms.
+ *
+ * Copyright (C) 2009 LSI Logic
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifdef CONFIG_ARCH_APP3K
+
+#include <mach/pmr.h>
+#include <asm/hardware/gic.h>
+
+	.macro	disable_fiq
+	.endm
+
+	.macro  get_irqnr_preamble, base, tmp
+	ldr	\base, =PMR_CII_BASE
+	.endm
+
+	.macro  arch_ret_to_user, tmp1, tmp2
+	.endm
+
+	.macro	get_irqnr_and_base, irqnr, irqstat, base, tmp
+	ldr     \irqstat, [\base, #GIC_CPU_INTACK]
+	ldr     \tmp, =1021
+	bic     \irqnr, \irqstat, #0x1c00
+	cmp     \irqnr, #29
+	it	cc
+	cmpcc   \irqnr, \irqnr
+	it	ne
+	cmpne   \irqnr, \tmp
+	it	cs
+	cmpcs   \irqnr, \irqnr
+	.endm
+
+	.macro test_for_ipi, irqnr, irqstat, base, tmp
+	bic     \irqnr, \irqstat, #0x1c00
+	cmp     \irqnr, #16
+	it	cc
+	strcc	\irqstat, [\base, #GIC_CPU_EOI]
+	it	cs
+	cmpcs	\irqnr, \irqnr
+	.endm
+
+	.macro test_for_ltirq, irqnr, irqstat, base, tmp
+	bic     \irqnr, \irqstat, #0x1c00
+	mov     \tmp, #0
+	cmp     \irqnr, #29
+	itt     eq
+	moveq   \tmp, #1
+	streq   \irqstat, [\base, #GIC_CPU_EOI]
+	cmp     \tmp, #0
+	.endm
+
+#endif
diff --git a/arch/arm/mach-app/include/mach/irqs.h b/arch/arm/mach-app/include/mach/irqs.h
new file mode 100644
index 0000000..901ef42
--- /dev/null
+++ b/arch/arm/mach-app/include/mach/irqs.h
@@ -0,0 +1,101 @@
+/*
+ *  include/mach/irqs.h
+ *
+ *  Copyright (C) 2004 Agere Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifndef __ASM_ARCH_IRQS_H
+#define __ASM_ARCH_IRQS_H
+
+#ifdef CONFIG_ARCH_APP3K
+/*
+ * Interrupt numbers
+ */
+#define INT_LOCALTIMER     29
+#define INT_LOCALWATCHDOG  30
+
+/* 32, unused */
+#define INT_SERIAL         33
+#define INT_I2C            34
+#define INT_SPI            35
+/* 36, unused */
+#define INT_ASI            37
+#define INT_GPIO_0         38
+#define INT_GPIO_1         39
+#define INT_CT             40
+#define INT_MAC_RX         41
+#define INT_MAC_TX         42
+#define INT_MAC_DMA        43
+/* 44, unused */
+#define INT_TIMER_5        45
+#define INT_TIMER_6        46
+#define INT_TIMER_7        47
+#define INT_TIMER_C        48
+#define INT_ECC_SINGLE     49
+#define INT_ECC_DUAL       50
+#define INT_NAND           51
+#define INT_USB            52
+#define INT_TDM            53
+#define INT_SLIC           54
+#define INT_PCI_ERR        55
+#define INT_PCI_ARM        56
+#define INT_GPDMA          57
+#define INT_GPDMA_ERR      58
+#define INT_GPDMA_TC       59
+#define INT_ASI_RXTX_0     60
+#define INT_ASI_RXTX_1     61
+/* 62, unused */
+/* 63, unused */
+
+#define NR_IRQS 64
+
+/*
+ *  Interrupt bit positions
+ */
+#define INTMASK_INT_SERIAL	(1 << (INT_SERIAL-32))
+#define INTMASK_INT_I2C		(1 << (INT_I2C-32))
+#define INTMASK_INT_SPI		(1 << (INT_SPI-32))
+#define INTMASK_INT_ASI		(1 << (INT_ASI-32))
+#define INTMASK_INT_GPIO_0	(1 << (INT_GPIO_0-32))
+#define INTMASK_INT_GPIO_1	(1 << (INT_GPIO_1-32))
+#define INTMASK_INT_CT		(1 << (INT_CT-32))
+#define INTMASK_INT_MAC_RX	(1 << (INT_MAC_RX-32))
+#define INTMASK_INT_MAC_TX	(1 << (INT_MAC_TX-32))
+#define INTMASK_INT_MAC_DMA	(1 << (INT_MAC_DMA-32))
+#define INTMASK_INT_TIMER_5	(1 << (INT_TIMER_5-32))
+#define INTMASK_INT_TIMER_6	(1 << (INT_TIMER_6-32))
+#define INTMASK_INT_TIMER_7	(1 << (INT_TIMER_7-32))
+#define INTMASK_INT_TIMER_C	(1 << (INT_TIMER_C-32))
+#define INTMASK_INT_ECC_SINGLE	(1 << (INT_ECC_SINGLE-32))
+#define INTMASK_INT_ECC_DUAL	(1 << (INT_ECC_DUAL-32))
+#define INTMASK_INT_NAND	(1 << (INT_NAND-32))
+#define INTMASK_INT_USB		(1 << (INT_USB-32))
+#define INTMASK_INT_TDM		(1 << (INT_TDM-32))
+#define INTMASK_INT_SLIC	(1 << (INT_SLIC-32))
+#define INTMASK_INT_PCI_ERR	(1 << (INT_PCI_ERR-32))
+#define INTMASK_INT_PCI_ARM	(1 << (INT_PCI_ARM-32))
+#define INTMASK_INT_GPDMA	(1 << (INT_GPDMA-32))
+#define INTMASK_INT_GPDMA_ERR	(1 << (INT_GPDMA_ERR-32))
+#define INTMASK_INT_GPDMA_TC	(1 << (INT_GPDMA_TC-32))
+#define INTMASK_INT_ASI_RXTX_0	(1 << (INT_ASI_RXTX_0-32))
+#define INTMASK_INT_ASI_RXTX_1	(1 << (INT_ASI_RXTX_1-32))
+
+#endif
+
+extern void app_init_irq_gic(void);
+
+#endif /*__ASM_ARCH_IRQS_H*/
diff --git a/arch/arm/mach-app/irq.c b/arch/arm/mach-app/irq.c
new file mode 100644
index 0000000..586fec4
--- /dev/null
+++ b/arch/arm/mach-app/irq.c
@@ -0,0 +1,29 @@
+/*
+ *  linux/arch/arm/mach-app/irq.c
+ *
+ *  Copyright (C) 2004 Agere Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+#include <linux/init.h>
+
+#include <asm/hardware/gic.h>
+#include <mach/hardware.h>
+
+void __init app_init_irq_gic(void)
+{
+	gic_dist_init(0, (void __iomem *)(APP_PMR_BASE + 0x1000), 29);
+	gic_cpu_init(0, (void __iomem *)(APP_PMR_BASE + 0x100));
+}
-- 
1.6.5.2

